### ISSUE TITLE
As a user, the edit + save experience in playground is consistent with other editors

### ISSUE THREAD

**Issue by EdMoffatt** [hyperledger/composer#1530]
Here's an attempt at articulating what we're after in a diagram.

It coins the following terms for referencing the 3 separate things that are needed to make it work.

**playground_model** - the current state of the files in the Define tab of Playground
**business_network_definition** - built from the playground_model for verification purposes: determines the state of the various error messages and action buttons
**deployed_business_network** - the latest business_network_definition that has been deployed (or updated) down to the Fabric

![image](https://user-images.githubusercontent.com/7454304/28061545-ddc5dbaa-6622-11e7-803f-3eb9ecaae013.png)

This is a suggested way of addressing #922 and #1498 

It should sort of remove the need for 922, because the fact **playground_model** is being stored as a separate thing means that we can still be testing against the **deployed_business_network** without throwing away what the user was working on in the Define tab when they navigate over to the Test tab.

**Comment by EdMoffatt** [comment]
Lets tag some people for input :)

@dselman - this is what we discussed yesterday morning.  Can you check to see if I've accurately represented what we discussed?

@nklincoln @jt-nti @samjsmith - we've all spoken about 922 or 1498 at some point or other, so would value your input on this suggestion too!  (It's very similar to some of the things you've previously suggested, but again I want to make sure I'm not mangling the idea in my issue writeup!)

**Comment by nklincoln** [comment]
There are not enough "like" emojis to reflect my thought on the above. We absolutely need to disassociate the playground 'in progress' edit with a complete compiled BND that can be deployed.

I like the division between working on plain text items that are validated in their context (with all changes -including errors - being persisted), with the deploy process being responsible for creating the BND from the (validated) plain text files and then performing the deploy.

**Comment by jt-nti** [comment]
@EdMoffatt does the logout warning mean that any undeployed changes will be thrown away? I'd prefer it if we could recover undeployed changes (the playground_model?), if there is one, in preference to retrieving the deployed_business_network when connecting/logging in.

**Comment by prasoc** [comment]
I have been working on being able to import broken BNA's (e.g with syntax errors) so that the end-user can work on the archive before deploying a functional BNA. This would have the same effect of splitting the "deployment" phase from the playground's current state, depending on what action you are trying to perform.

The way I've implemented it is to hook into the import filetypes' "deploy" functions, like this:

packages/composer-common/lib/aclmanager.js:
```
   /**
     * Set the AclFile for this AclManager
     * @param {AclFile} aclFile  - the AclFile to associate with this AclManager
     * @param {bool} [validate=true] - Determine whether the file should be validated
     * @private
     */
    setAclFile(aclFile, validate) {
        validate = validate !== false; // default initializer

        if(validate) {
            aclFile.validate();
        }

        this.aclFile = aclFile;
    }
```

and the corresponding functionality change for each filetype. This allows you to store broken files within the network definition, just need to do a check when pressing the "Deploy" button to see if the validation passes. The changes here would be just the same as when it auto-saves: you ALWAYS save the file to the defintion no matter how broken it is, and let the user sort it out (by disabling the Deploy functionality)

So for instance, in [my branch here](https://github.com/prasoc/composer/tree/import-dialogs), you want to import an archive, it first tries to import it with the validation routines enabled and if that fails, informs the user that the archive has some errors. 
If they want, they can import it anyway which skips each filetype's ".validate()" routine by passing `validate = false` to the code above.

I am still very new to contributing here,, so I would love some critiques on my implementation for solving this limitation.

**Comment by EdMoffatt** [comment]
@jt-nti - I'm not sure...  Would we store a playground_model for all the business networks that you can access?  e.g. if I'm at a stage where I'm working on a bunch of them, I'll then need to be potentially storing playground_models for all of them.  Just needs more storage :)

There needs to be a way that I say "go back to the underlying Fabric and load the latest deployed business network into my UI instead of whatever I was working on before".  In my head, "log off and then re-connect" seemed a pretty neat way of achieving that, but happy to throw around other ideas :)

**Comment by EdMoffatt** [comment]
@prasoc - that's an interesting idea :)  I'm the *worst* person to comment on implementation choices though, so I'll leave it to others to weigh in!

**Comment by samjsmith** [comment]
@EdMoffatt It looks to me like the 'sensible trigger' you mentioned is any codeChanged event and is debounced to only trigger every 700ms (although I can't find where this is handled in the CodeMirror source at the moment).

In terms of the behaviour you've specified, it looks good to me.  We've spoken of having a 'deployed/undeployed' state internally in the UI and I think that's what we're getting at changing here.

As far as I can see, the only part that's missing from your diagram ie. what happens when the user has edited the model, it's been validated and, regardless of the result, the user then wants to just nope out and click away.

ie. we need an `You're about to discard changes OK/Cancel` modal here I think?

I like the idea of the 'Reset to what's deployed' idea too (with associated levels of 'are you sure? are you really really sure?').  That sounds like a new button to me so that could be done in a separate item.

I've skimmed over @prasoc commit there which permits the import of invalid BNAs and it looks good in principle to me (although I haven't actually tried it and I think there's at least one test missing).  It changes the common component (in a good way) to allow validation to be disabled (but default to being enabled) but it probably needs @sstone1 to cast his architectural eye over it too. 

So I'd suggest we could also commit this under a separate item of `allow import of invalid BNA`

In summary, here's what I thought:
This issue -> sort out persistence of files in editor with new modal and enabling/disabling of Deploy/Export buttons
Associated New Issue 1 -> Re-load playground with what's deployed button/flow
Associated New Issue 2 -> Integrate @prasoc commit to allow import of invalid BNA archives to complete the picture.

This issue probably needs a good blockchain triage to arrive at a consensus!

**Comment by sstone1** [comment]
The diagram at the top looks great to me. I fully endorse separating the users in-progress project from what is actually deployed to the blockchain, and it looks like we have a way to make that make sense to users.

This user story is also relevant when we do the work in this space:
As a user, I can import an invalid BNA and rectify the errors from the playground #1543

**Comment by samjsmith** [comment]
@prasoc do you want to submit a pull request for your commit that we can review and integrate?

**Comment by prasoc** [comment]
@samjsmith yes, I will implement some tests and get the last few things sorted and pull the changes in this weekend.

**Comment by prasoc** [comment]
Hi all,
Take a look at pull request #1722 and see my implementation. I am looking for comments, and to continue the discussion on the way forward.

It's a very simple way which isn't a "fully-featured" fix just yet: BNAs which throw an IllegalModelException are able to be imported, but not ALL broken archives

**Comment by EdMoffatt** [comment]
Suggested first-step bug fix of this (nicked from #2309 ):

* * *

Allow a user to switch between files in the Define tab without any that are in error state being "reverted" to older versions. This probably means we need to store the latest edits to all files in memory.

We will go further with this fix, but for now this seems a good minimal required to fix the worst of the bug. So to be explicit...

If the user goes back to the My Business Networks screen, any un-deployed changes will still be lost.
If the user switches to any other tab (e.g. Test panel), any un-deployed changes will still be lost.
Obviously there are better versions of these behaviours, and that's what we shall explore in the Epic - for now, lets just try to make the basics of writing a Transaction Processor Function workable ;)

* * *

Also adding the bug tag, because getting at least this part of the behaviour fixed should definitely be a priority bug for us!

**Comment by caroline-church** [comment]
done so closing
